tools/redhat/osv.py

```diff
@@ -0,0 +1,208 @@
+        self.related: list[str] = []
+
+        if published:
+            self.published = published
+        else:
+            self.published = modified
+        self.modified = modified
+
+        self.summary = csaf_data.title
+
+        # Set severity to the CVSS of the highest CVSSv3 base score
+        vulnerability_scores: dict[str, str] = {}
+        for vulnerability in csaf_data.vulnerabilities:
+            if not vulnerability.cvss_v3_vector or not vulnerability.cvss_v3_base_score:
+                continue
+            vulnerability_scores[
+                vulnerability.cvss_v3_base_score] = vulnerability.cvss_v3_vector
+        if vulnerability_scores:
+            highest_score = sorted(vulnerability_scores.keys())[-1]
+            self.severity = [{
+                "type": "CVSS_V3",
+                "score": vulnerability_scores[highest_score]
+            }]
+
+        self.affected: list[Affected] = []
+        for vulnerability in csaf_data.vulnerabilities:
+            self.related.append(vulnerability.cve_id)
+            for remediation in vulnerability.remediations:
+                self.affected.append(Affected(remediation))
+
+        self.references = self._convert_references(csaf_data)
+
+    def _convert_references(self, csaf) -> list[dict[str, str]]:
+        """
+        CSAF has references for an advisory and each vulnerability has references as well.
+        Collect this into a single references list for OSV and deduplicate them.
+        """
+        references: dict[str, str] = {}
+        for reference in csaf.references:
+            # This will capture both the Advisory URL and the CSAF document for the advisory
+            if reference["category"] == "self":
+                if reference["summary"].startswith(REDHAT_ADVISORY_URL):
+                    self.id = reference["summary"].removeprefix(
+                        REDHAT_ADVISORY_URL)
+                references[reference["url"]] = "ADVISORY"
+            else:
+                references[reference["url"]] = self._get_reference_type(
+                    reference)
+        for vulnerability in csaf.vulnerabilities:
+            for reference in vulnerability.references:
+                # This captures the CVE specific information
+                if reference["category"] == "self":
+                    references[reference["url"]] = "REPORT"
+                else:
+                    references[reference["url"]] = self._get_reference_type(
+                        reference)
+        return [{"type": t, "url": u} for u, t in references.items()]
+
+    def _get_reference_type(self, reference: dict[str, str]) -> str:
+        """
+        Convert references from CSAF into typed referenced in OSV
+        Also make sure to add a related entry for any GO advisory references found
+        """
+        if reference["url"].startswith(ADVISORY_URL_PREFIXES):
+            self._add_go_related(reference["url"])
```

Comment:
This reads a bit awkwardly given it is called for non-Go advisories too.

```suggestion
            self._maybe_add_go_related(reference["url"])
```

Reply:
Or maybe just pull out the conditional behaviour to the call site rather than inside the function?

Reply:
I removed this function and moved the code back in-line. I also renamed the calling function from `get_reference_type` to `get_reference_type_and_add_go_related` to better indicate it's dual behaviour.